### ISSUE TITLE
found_dags is None

### ISSUE THREAD

**Issue by vansivallab** [apache/airflow#899]
There seems to be an issue when airflow fails to import a dag, found_dags is None:
[2016-01-21 10:58:00,275] {models.py:204} INFO - Importing /Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/airflow/example_dags/example_bash_operator.py
[2016-01-21 10:58:00,275] {models.py:210} ERROR - Failed to import: /Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/airflow/example_dags/example_bash_operator.py
Traceback (most recent call last):
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/airflow/models.py", line 207, in process_file
    with utils.timeout(30):
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/airflow/utils.py", line 620, in **enter**
    signal.signal(signal.SIGALRM, self.handle_timeout)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/signal.py", line 47, in signal
    handler = _signal.signal(_enum_to_int(signalnum), _enum_to_int(handler))
ValueError: signal only works in main thread
[2016-01-21 10:58:00,282] {_internal.py:87} INFO - 127.0.0.1 - - [21/Jan/2016 10:58:00] "GET /admin/airflow/graph?dag_id=example_bash_operator HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask_admin/base.py", line 68, in inner
    return self._run_view(f, _args, *_kwargs)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask_admin/base.py", line 367, in _run_view
    return fn(self, _args, *_kwargs)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/flask_login.py", line 755, in decorated_view
    return func(_args, *_kwargs)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/airflow/www/utils.py", line 198, in view_func
    return f(_args, *_kwargs)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/airflow/www/utils.py", line 103, in wrapper
    return f(_args, *_kwargs)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/airflow/www/views.py", line 1203, in graph
    dag = dagbag.get_dag(dag_id)
  File "/Users/vansivallabhaneni/.pyenv/versions/3.5.0/lib/python3.5/site-packages/airflow/models.py", line 171, in get_dag
    if dag_id in [dag.dag_id for dag in found_dags]:
TypeError: 'NoneType' object is not iterable

I think when handling the exception (model.py line 209), it should return an empty list not None.

**Comment by vansivallab** [comment]
And when I modify it to return an empty list, I get a error saying that the dag is missing, which is odd because it has to load it to render the /admin page.

**Comment by r39132** [comment]
Hi!
Can you provide some details here such as 
1. Version of code (either release version or `git log | head` output if you are running a different version
2.  A sample of your DAG
3. Steps to reproduce
4. Deployment architecture (e.g. single node with both Webserver and Scheduler+LocalExecutor running, Celery installation, etc....?)
5. Any other details

**Comment by r39132** [comment]
Closing for lack of info.

**Comment by abridgett** [comment]
I'm also getting this ATM.  
1. (I realise we are a bit behind master, will probably wait for a release if one is due)

```
commit 0ee4eb8ee0a39787b377a78264dd2ff9b14959f9
Author: Adrian Bridgett <adrian@smop.co.uk>
Date:   Fri Jan 15 14:43:13 2016 +0000

    revert hotfix - it breaks the dialog box

commit 3f5ccfb94dd29318c53395b0a47394543f7d185a
Author: Adrian Bridgett <adrian@smop.co.uk>
Date:   Fri Jan 15 14:41:50 2016 +0000

    state wasn't being saved

commit ccf30cd36fa79b12f21b76c5c40ab21dbc8568ac
Merge: b299a23 459ec4c
Author: Adrian Bridgett <adrian@smop.co.uk>
Date:   Fri Jan 15 14:42:37 2016 +0000

    Merge branch 'master' into feature/os_prod
```
1. Not sure I can ATM, happy to share privately
2. This was a new DAG, copied it over, restarted webserver. Error in my code, repeated a few times.  Finally it tried to run, found a bug in one of the tasks.  Clicking on anything related to the dag (tree view etc) in the UI causes the same error.  Clicking on the (i) info button does work however. 
3. Linux with Celery
4. Dag runs list one run (failed)

**Comment by abridgett** [comment]
Hmm - restarting scheduler and then webserver fixed it.  Our deploy is restarting the webserver anyway immediately after the new dags are copied.

**Comment by abridgett** [comment]
I think I've got a clue - in the webserver logs process_file is aborting on an exception, so it's returning None rather than found_dags (my line numbers will be a bit off due to debug)

```
Traceback (most recent call last):
  File "/opt/airflow/lib/python3.4/site-packages/airflow/models.py", line 213, in process_file
    with utils.timeout(30):
  File "/opt/airflow/lib/python3.4/site-packages/airflow/utils.py", line 620, in __enter__
    signal.signal(signal.SIGALRM, self.handle_timeout)
ValueError: signal only works in main thread
```

Removing the `with utils.timeout(30):` wrapper from process_file fixes this error but I'm sure it's not the best approach!

**Comment by mistercrunch** [comment]
This bug should be fixed in `master`. I added try statements inside the timeout context manager so that it goes disabled when using LocalExecutor or other subprocesses. 

This timeout is important as it makes sure that if a pipeline takes a long time to parse (say someone would run a Hive job in the python code of the pipeline instead of using an operator), the script times out after 30 seconds and all the other pipelines can move on. Without the timeout, a bad actor could hold up the scheduler for long periods.

We may want to take a different, more generic approach to timeouts eventually, but I the current approach should be good. Closing.
